### ISSUE TITLE
WCAG Text & Link with a predefined background color

### ISSUE THREAD

**Issue by michaelbourne** [themeum/kirki#2267]
### Issue description:

Looking at the docs, it tells me the background color used for reference is a control. What if I wanted all backgrounds to be #fff, can i code that in instead of a color control?

**Comment by aristath** [comment]
This has never been requested before so right now it's not an option. 
However, it would be fairly simply to implement it and release an update for those controls if you require that feature. 
We use the background-color control to add a listener and change values depending on the color selected in the background.
Defining a hardcoded color would simplify things a lot, so if you require that it can be done.

**Comment by michaelbourne** [comment]
That's great to know. I would love an update with that (or even a pointer to do it myself if you chose not to make it an official feature). Let me know if you prefer corresponding here or email after my purchase.

**Comment by aristath** [comment]
This was implemented for the wcag-link-color control.
I don't know though if it would make sense for the wcag-text-color control... I mean if the background-color is hardcoded, then there's nothing to query and watch, it will be a simple array of accessible colors that will never change because there won't be anything to listen to for changes.

Could you please elaborate a bit on how you envisioned that would work? If I understand what it is you're looking for I'm sure we can accommodate that scenario.

I want to run some more tests for the link-color control and I'll update the version on the site tomorrow if there's no issues 👍

**Comment by michaelbourne** [comment]
Hey @aristath 

I thought about it some more myself with regards to how I want to use it. I'm trying to think of ways to use the controls outside of the box.

In my case, I will always want a white background on the site. So having the option to hard code that into either control would be great.

My first thought: When it comes to the text color control, it would simply prevent them from picking a color that doesn't have enough contrast against white. The control may be called "text color", but realistically I can use the color anywhere I want, so it could instead be used as a "CTA color" or "brand color" in my theme. In that case, I could grab the color selected, drop the HSL lightness to 0 to have a hue matched black for the body font, and a brand color to use for highlights and CTAs that I know will pass WCAG scans.

The alternative would be a wcag-link-color with a hardcoded background and a hardcoded text color of my choice, and using the wcag-link-color output in a similar way (links, buttons, highlights, etc).

Specific use case, the industry I'm building for has a horrible habit of gold-on-white (for links, banners, buttons, etc), which fails every level of WCAG. I'm trying to prevent that.

**Comment by aristath** [comment]
> My first thought: When it comes to the text color control, it would simply prevent them from picking a color that doesn't have enough contrast against white. The control may be called "text color", but realistically I can use the color anywhere I want, so it could instead be used as a "CTA color" or "brand color" in my theme.

Hmmmm OK... Yeah, in that case we can add the option to hardcode a background-color value in the control. I'll take care of it tomorrow and update both controls on the site. 👍 

But if you just want an array of colors for that, you could also just use something else... Just throwing some ideas here, but take a look for example at https://casesandberg.github.io/react-color/
The `Github`, `Twitter` (with the custom-color field hidden) and the `Circle` pickers could very well be used to add a "color-palette" with a few hardcoded text-colors that you know are accessible. Similar concept with the wcag control - but hardcoded instead of dynamic.
So if all you want is a palette, I already have a control that uses react-color (see https://github.com/kirki-framework/control-react-color/) and that can also get the job done.

> Specific use case, the industry I'm building for has a horrible habit of gold-on-white

wait... what? WTH????
Well, at least it's not red-on-green like I've seen some sites use on Christmas. 🤣 

Both controls will be uploaded tomorrow. 👍

**Comment by michaelbourne** [comment]
You rock :)

As a dev I much prefer the react route, but my design lead insists that users be allowed to choose any color they want, so long as it's WCAG compliant for contrast (please excuse me if I wildly misunderstood how the "custom" tab on your control worked, i assumed it still ran WCAG checks)

**Comment by aristath** [comment]
> but my design lead insists that users be allowed to choose any color they want, so long as it's WCAG compliant for contrast

Well, you _could_ tweak the arguments passed-on to the control... So for example you could define smaller steps.

The defaults are:
```
'stepHue'             => 15,
'maxSaturation'       => 0.5,
'stepSaturation'      => 0.1,
'stepLightness'       => 0.05,
'contrastThreshold'   => 4.5
```

Which means that for a white background, it would do the following:
1. Create an array of hues using `stepHue` increments, so `360 / 15` = **24** hues)
2. Check saturation levels from `0` to `maxSaturation` in `stepSaturation` increments - so `0.5 / 0.1` = **5**.
3. Check lighness levels in `stepLightness` increments, and since the background is white it will check between `0` and `0.5` - so `0.5 / 0.05` = **10**

Put that all together, and it checks a total of `24 * 5 * 10` = 1200 colors, then checks the contrast of each one with the background color, and returns the ones that have a contrast higher than the value defined in `contrastThreshold`.

The reason I'm using the above defaults is because these queries have to run every single time a color changes. So imagine having a background-color colorpicker, and dragging the color from white to black to colors and so on. It has to be **fast**, so I'm limiting the query to _only_ 1200 colors.
However, since in your case there will not be a background-color control you can go wild with the values and have it query colors using a lot smaller steps. This will produce a more detailed palette so users can pick an accessible color from a wider array. Performance won't be an issue in your case - unless you have 20 controls like that, each one querying 100.000 colors when the customizer loads. Then yes, you can expect the customizer's initial load to be 10s.

> please excuse me if I wildly misunderstood how the "custom" tab on your control worked, i assumed it still ran WCAG checks

Yeah, `custom` is completely unfiltered... no checks run there. 

I'm currently in the process of rewriting these controls with React and I'm changing the implementations to be prettier, faster and overall awesome(er). And there _will_ be checks done in the custom control, showing the rating & contrast to users. But it won't limit them, they'll still be able to choose something completely inaccessible if they so desire... but at least they'll know that they made a horrible choice.

I don't have an ETA for that though, React in the customizer is a bit temperamental.
